### PR TITLE
Lecture 11: Kalman Filtering and Topic Models

### PR DESCRIPTION
**Pull request by DBay-ani**
Link to rendered page for TAs to review:
https://dbay-ani.github.io/pgm-spring-2019/notes/lecture-notes-template/

### CODE REVIEW

File: _posts/2019-02-20-lecture-notes-template.md [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+authors:
+  - name: Aakanksha Naik
+    url: "#"  # optional URL to the author's homepage
+  - name: D. Bayani
+    url: "#"
+  - name: Bingqing Chen
+    url: "#"
+
+editors:
+  - name: Editor 1  # editor's full name
```
Comment (pliang279) [inline _posts/2019-02-20-lecture-notes-template.md]: Editor: Paul Liang

File: _posts/2019-02-20-lecture-notes-template.md [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+V_{t+1} &=
+  \begin{bmatrix}
+    P_{t+1|t} & P_{t+1|t}C^T \\
+    CP_{t+1|t} & CP_{t+1|t}C^T + R
+  \end{bmatrix}
+\end{aligned}
+</d-math>
+
+Given a joint distribution of two Gaussian random variables $X_1, X_2$, recall that their conditional and marginal probability distributions are defined as follows:\\
+$$ P(X_2) = \mathcal{N}(M_2^m, V_2^m) $$ \\
```
Comment (pliang279) [inline _posts/2019-02-20-lecture-notes-template.md]: please use \begin{align} and center the equations

File: _posts/2019-02-20-lecture-notes-template.md [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+\begin{aligned}
+\hat X_{t+1 \vert t+1} &= \hat X_{t+1 \vert t} + K_{t+1} (Y_{t+1} - C \hat X_{t+1 \vert t}) \\
+P_{t+1 \vert t+1} &= P_{t+1 \vert t} - KCP_{t+1 \vert t}\\
+\end{aligned}
+</d-math>
+Here $K_{t+1} = P_{t+1 \vert t} C^T (C P_{t+1 \vert t} C^T + R)^{-1}$ and is known as the Kalman gain matrix. An interesting point to note is that $K_t$ is independent of $Y_t$ (i.e. the data), so it can be precomputed to improve efficiency.
+
+### Example of Kalman Filtering in 1-D
+Suppose we have some noisy observations of a particle performing a random walk in 1-D. Assume our states and observations are defined as follows: \\
+$ X_{t \vert t-1} = X_{t-1} + W $ where $W \sim \mathcal{N}(0,\sigma_x)$ \\
```
Comment (pliang279) [inline _posts/2019-02-20-lecture-notes-template.md:137]: please use \begin{align} and center the equations

File: _posts/2019-02-20-lecture-notes-template.md [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+matrices, we must do learning. Using approaches like EM we can interleave learning and inference
+to come across the parameters of interest.
+
+Furthering this comparison to HMMs, the Rauch-Tung-Strievel algorithm allows us to perform "exact off-line inference in an LDS", and is essentially a "Guassian analog of the forwards-backwards" algorithm.
+
+## Learning SSMs
+In order to learn the necessary parameters for the Kalmann filter, 
+we calculate the complete data likelihood:
+$$l_t(\theta, D) = \sum_{n}p(x_n, y_n) = $$
+$$(\sum_n log(p(x_1))) + (\sum_n\sum_t log(p(x_{n,t} | x_{n, t-1})) + (\sum_n\sum_t log(p(y_{n,t} | x_{n, t}))) =$$
```
Comment (pliang279) [inline _posts/2019-02-20-lecture-notes-template.md]: use \log

File: _posts/2019-02-20-lecture-notes-template.md [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+$\hat X_{t+1 \vert t+1} &= \hat X_{t+1 \vert t} + K_{t+1} (Z_{t+1} - C \hat X_{t+1 \vert t}) = \frac{(\sigma_t + \sigma_x)Z_t + \sigma_z \hat X_{t|t}}{\sigma_t + \sigma_x + \sigma_z}$
+$P_{t+1 \vert t+1} = \frac{(\sigma_t + \sigma_x) \sigma_z}{\sigma_t + \sigma_x + \sigma_z}$
+
+### Understanding the intuition behind Kalman Filtering
+In the KF update equation for the mean, $\hat X_{t+1 \vert t+1} &= \hat X_{t+1 \vert t} + K_{t+1} (Z_{t+1} - C \hat X_{t+1 \vert t})$, the term $(Z_{t+1} - C \hat X_{t+1 \vert t})$ is called the **innovation** term. We can see that the update equation for new belief is a convex weighted combination of updates from prior and observation, with the Kalman Gain matrix acting as the weight. From the equation for the Kalman Gain matrix, we can see that if observations are noisy ($\sigma_z$ or $R$ is large), then the KG matrix is small and updates rely more on prior. On the other hand if the process is unpredictable (large $\sigma_x$) or prior is unreliable (large $\sigma_t$), the KG matrix is higher and we rely more on the observation. 
+
+
+## Discussion of Where the A, G, and C Matrix Come From
```
Comment (pliang279) [inline _posts/2019-02-20-lecture-notes-template.md]: "Derivation of A, G, and C Matrices"

File: _posts/2019-02-20-lecture-notes-template.md [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+Thus far in the course, we have covered the elimination algorithm, message passing, and
+algorithms that are powered based on those principles. These techniques allowed for 
+exact inference, and while they have been shown useful, they cannot cope with many of 
+the settings and problems we wish to explore. This motivates uses of approximate
+techniques, which we will begin discussing in this lecture.
+
+
+
+
+## Some Discussion of a Task, and How to Tackle It with Appropriate Modeling
```
Comment (pliang279) [inline _posts/2019-02-20-lecture-notes-template.md]: "Appropriately Modeling Different Tasks"

File: _posts/2019-02-20-lecture-notes-template.md [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+## Discussion of Where the A, G, and C Matrix Come From 
+
+
+Up to this point, we have discussed inference in the Kalmann filter model; given the 
+model up-front, tell me something about the data. This leaves open where the matrices A, G, and
+C come from, however. This is a similar situation we were in for HMMs: to find the necessary 
+matrices, we must do learning. Using approaches like EM we can interleave learning and inference
+to come across the parameters of interest.
+
+Furthering this comparison to HMMs, the Rauch-Tung-Strievel algorithm allows us to perform "exact off-line inference in an LDS", and is essentially a "Guassian analog of the forwards-backwards" algorithm.
```
Comment (pliang279) [inline _posts/2019-02-20-lecture-notes-template.md]: please add more details here perhaps a citation to a paper

File: _posts/2019-02-20-lecture-notes-template.md [hunk 8]
```diff
@@ -0,0 +1,361 @@
...
+
+## Learning SSMs
```
Comment (pliang279) [inline _posts/2019-02-20-lecture-notes-template.md]: "Learning State Space Models"

File: _posts/2019-02-20-lecture-notes-template.md [hunk 9]
```diff
@@ -0,0 +1,361 @@
...
+
+
+
+## Mean-Field Assumption (In Topic Models)
+
+Recall the form of the true posterior :
+
+$$p(\beta, \theta, z | w) = \frac{p(\beta, \theta, z, w)}{p(w)}$$
+
+Suppose that in q (our approximation to p) we could break dependancies in the jiont by assuming
```
Comment (pliang279) [inline _posts/2019-02-20-lecture-notes-template.md]: $q$, $p$, jiont is a typo

File: _posts/2019-02-20-lecture-notes-template.md [hunk 10]
```diff
@@ -0,0 +1,361 @@
...
+
+Notice that in this fully-factored model, each factor is a term of a distribution - 
+this is unlike the factorization we say over Bayes Nets before, where conditional 
+distributions appear. As such, if we need to answer a query of form $$q(z_d)$$, 
+marginalizing across $$q(\beta, \theta, z)$$ is trivial since we now that moving the
+sums into the product, $$\sum_{\beta, \theta}\Pi_{k}q(\beta_k)\Pi_{d}q(\theta_d) = 1$$.
+
+In general with variational methods, the true posterior might not exist in the 
+family of $$q$$ we consider - we buy tractability with the cost of approximation error.
+
```
Comment (pliang279) [inline _posts/2019-02-20-lecture-notes-template.md:376]: please go through and proof-read one more time!

File: _posts/2019-02-20-lecture-notes-template.md [hunk 11]
```diff
@@ -0,0 +1,361 @@
...
+$$log(p(x)) = 
+KL(q_{\theta}(z | x) || p_{\theta}(z | x)) + \int_z q_{\theta}(z | x)log\frac{p_{\theta}(x, z)}{q_{\theta}(z | x)}dz 
+\ge \int_z q_{\theta}(z | x)log\frac{p_{\theta}(x, z)}{q_{\theta}(z | x)}dz
+:= \mathscr{L}(\theta, \phi ; x)$$
+
+Equivalently, we can minimize the aforementioned "free-energy" of the system:
+
+
+$$F(\theta, \phi ; x) = -log(p(x)) + KL(q_{\theta}(z | x) || p_{\theta}(z | x))$$
```
Comment (pliang279) [inline _posts/2019-02-20-lecture-notes-template.md]: \log

File: _posts/2019-02-20-lecture-notes-template.md [hunk 12]
```diff
@@ -0,0 +1,361 @@
...
+
+
+We start by introducing another distribution , $Q_{\theta}$ which is a distribution on the 
+same set of random variables. The goal, eventually, is to have $Q_{\theta}$ be a distribution
+that is easier to work with than $p$ but is sufficiently "close" to serve as a reasonable 
+surrogate in calculations. 
+
+Using this, we want to maximize the lower bound for the log-likelihood:
+
+$$log(p(x)) =
```
Comment (pliang279) [inline _posts/2019-02-20-lecture-notes-template.md]: \log

**Review by pliang279 — CHANGES_REQUESTED** [review]

**Comment by pliang279** [comment]
I have given some comments, please make these edits and proof read the notes, thanks!

File: _posts/2019-02-20-lecture-notes-template.md (hunk 1 above)
Comment (DBay-ani) [reply _posts/2019-02-20-lecture-notes-template.md]: done

File: _posts/2019-02-20-lecture-notes-template.md (hunk 2 above)
Comment (DBay-ani) [reply _posts/2019-02-20-lecture-notes-template.md]: done

File: _posts/2019-02-20-lecture-notes-template.md (hunk 3 above)
Comment (DBay-ani) [reply _posts/2019-02-20-lecture-notes-template.md:137]: done

File: _posts/2019-02-20-lecture-notes-template.md (hunk 4 above)
Comment (DBay-ani) [reply _posts/2019-02-20-lecture-notes-template.md]: done

File: _posts/2019-02-20-lecture-notes-template.md (hunk 5 above)
Comment (DBay-ani) [reply _posts/2019-02-20-lecture-notes-template.md]: See the recently pushed edits. This section was at best a digression on the part of Dr. Xing as part of answering a student's question, and took only about 2 minutes of class time. As such, it does not warrant a title indicating a derivation was performing, when nothing but prose was offered in the original material.

File: _posts/2019-02-20-lecture-notes-template.md (hunk 8 above)
Comment (DBay-ani) [reply _posts/2019-02-20-lecture-notes-template.md]: done

File: _posts/2019-02-20-lecture-notes-template.md (hunk 9 above)
Comment (DBay-ani) [reply _posts/2019-02-20-lecture-notes-template.md]: I am not sure what you mean by this. This is exactly the form provided by Dr. Xing. Q was not conditional here.

File: _posts/2019-02-20-lecture-notes-template.md (hunk 10 above)
Comment (DBay-ani) [reply _posts/2019-02-20-lecture-notes-template.md:376]: I added some clarification in regards of not needing to marginalize on a Q with the given form. I hope that this addresses the concern you had; I presume the wording provided previously suggested to you that marginalization would still be necessary.

File: _posts/2019-02-20-lecture-notes-template.md (hunk 11 above)
Comment (DBay-ani) [reply _posts/2019-02-20-lecture-notes-template.md]: done

File: _posts/2019-02-20-lecture-notes-template.md (hunk 12 above)
Comment (DBay-ani) [reply _posts/2019-02-20-lecture-notes-template.md]: done

**Comment by DBay-ani** [comment]
Pushed changes that we believe address your concerns ; please let us know if not.

**Comment by DBay-ani** [comment]
Also, do you receive emails when a person leaves a comment like this? I assume so, but given the time-frame, I might reach out in email to be sure if a few hours pass - that is, to be sure I do due-diligence in making you aware that we are prepared for your re-review.

File: _posts/2019-02-20-lecture-notes-template.md (hunk 6 above)
Comment (DBay-ani) [reply _posts/2019-02-20-lecture-notes-template.md]: Done. Saw this comment in the email I received and addressed it then, but did not see it here earlier.

File: _posts/2019-02-20-lecture-notes-template.md (hunk 7 above)
Comment (DBay-ani) [reply _posts/2019-02-20-lecture-notes-template.md]: Saw this comment in the email I received and addressed it then, but did not see it here earlier.

Please see the comment made above regarding "
Derivation of A, G, and C Matrices"
As stated there, the content from Dr. Xing there was intentionally high-level to begin with, and provided primarily to shape the spirit of the lecture as opposed to introduce substantive material.

**Comment by alshedivat** [comment]
I believe the merged version of these notes is incorrectly titled as Lecture 10 instead of 11. Also, all other lectures are titled `Lecture xx: <lecture name>`, but this one just states lecture number.

@pliang279, since this has been already merged, could you push title fixes directly into `master`? Thanks!

**Comment by DBay-ani** [comment]
To comment on this:
(1) The title: this was a point of some difficulty in terms of guessing what the TAs might prefer. As noted in the beginning of the scribe notes, the actual lecture content differed substantially from what the title on the course website would entail. The best name for the lecture can probably be scrapped from the description field immediately below the title. Note, however, that that will be different than what is on the course website, and for good reason.
(2) Lecture number: this again is something that is best resolved by the TAs. Our lecture was 11th according to the course website, but the 10th one actually delivered due to the snow cancellation in January. Further, the video-sharing cite lists this as lecture 10.

**Comment by alshedivat** [comment]
@DBay-ani, thanks for explaining!

(1) Makes sense. Moving the description to the title and the note from the beginning of the lecture into the description would work best.

(2) Let's stick with the consistent numbering as given in the schedule, even though the content largely overlapped between the two lectures. As for lecture numbering in the videos, we will ask the videographer to fix that as well.

@DBay-ani, could you send another PR that addresses (1) and (2)? Thanks!

**Comment by DBay-ani** [comment]
I will try that ASAP. I had already deleted my fork.

**Pull request merged**